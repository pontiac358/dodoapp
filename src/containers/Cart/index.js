import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux'
import { getUniqProducts } from '../../helper/selector'
import { removeLineCart, clearCart, toOrder } from '../../actions/cart'
import styled from '@emotion/styled'
import CartItem from './components/CartItem'
import Btn from '../../shared/Btn'

const Container = styled.div`
    max-width:800px;
    padding: 0 10px;
    margin:auto;
`;

const ContainerBottom = styled.div`
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    align-items: center;
    padding: 20px 0;
  
`;

const TotalPrice = styled.div`
    margin-left:auto;
    font-size: 32px;
    @media only screen and (max-width: 480px) {
        font-size:20px;
        text-align:center;
        margin: 0;
        width: 100%;  
    }
`;

const ButtonWrap = styled.div`
    margin-left:auto;
    font-size: 32px;
    @media only screen and (max-width: 480px) {
       margin:auto;  
    }
`;

const EmptyCartHead = styled.h2`
    text-align: center;
    font-size: 32px;
    padding: 20px 0;
`;
EmptyCartHead.displayName = 'EmptyCartHead';

const ButtonWrapCenter = styled.div`
    text-align:center;
`;


export class Cart extends React.Component {
    handleRemoveLineCart = (id) => {
        removeLineCart(id)
    };

    handleClearCart = () => {
        clearCart()
    };

    handleToOrder = (e) => {
        alert('Заказ принят');
        toOrder();
        this.props.history.push('/')
    };

    render() {
        const { products, cart } = this.props;

        if(!products.length){
            return (
                <Container>
                    <EmptyCartHead>Корзина пуста</EmptyCartHead>
                    <ButtonWrapCenter>
                        <Btn to={'/'} title={'На главную'}/>
                    </ButtonWrapCenter>
                </Container>
            )
        }

        return (
            <Container>
                <ContainerBottom>
                    <ButtonWrap>
                        <Btn onClick={ this.handleClearCart } title={'Очистить корзину'}/>
                    </ButtonWrap>
                </ContainerBottom>
                {
                    products && products.map(item => {
                        return(
                            <CartItem removeLineCart={ this.handleRemoveLineCart } key={ item.product.id } { ...item }/>
                        )
                    })
                }
                <hr/>

                <ContainerBottom>
                    <TotalPrice>
                        Сумма заказа: { cart.totalPrice } р
                    </TotalPrice>
                </ContainerBottom>
                <ContainerBottom>
                    <ButtonWrap>
                        <Btn onClick={ this.handleToOrder } title={'Оформить заказ'}/>
                    </ButtonWrap>
                </ContainerBottom>
            </Container>
        );
    }
}

Cart.propTypes = {
    products: PropTypes.array,
    cart: PropTypes.object
};

const mapStateToProps = (state) => {
    return {
        cart: state.cart,
        products: getUniqProducts(state),
    }
};

export default connect(mapStateToProps)(Cart)