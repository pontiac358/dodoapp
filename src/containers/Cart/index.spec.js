import React from 'react'
import { shallow } from 'enzyme'
import { Cart }   from '../Cart'

describe('Cart container', () => {
    const props = {
        products: [],
        cart: {}
    };

    const wrapper = shallow(<Cart {...props} />);
    it("render correctly Cart container", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("if Cart empty", () => {
        expect(wrapper.find('EmptyCartHead').at(0).text()).toEqual('Корзина пуста');
    });

    describe('If Cart not empty', () => {
        const nextProps ={
            products: [{
                count: 1,
                totalPrice: 100,
                product: {
                    id: 1,
                    title: 'test',
                    price: 100
                }
            }],
            cart: {
                totalPrice: 100,
                count: 1,
                products:[]
            }
        };
        const nextWrapper = shallow(<Cart {...nextProps} />);

        it('renders <EmptyCartHead /> wrapper', () => {
            expect(nextWrapper.find('EmptyCartHead')).toHaveLength(0)
        });

        it('renders <CartItem /> wrapper', () => {
            expect(wrapper.find('CartItem')).toHaveLength(props.products.length)
        })
    });
});

