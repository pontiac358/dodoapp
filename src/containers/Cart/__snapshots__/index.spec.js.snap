// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart container render correctly Cart container 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Cart
    cart={Object {}}
    products={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef(render)>
          Корзина пуста
        </ForwardRef(render)>,
        <ForwardRef(render)>
          <Btn
            title="На главную"
            to="/"
          />
        </ForwardRef(render)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Корзина пуста",
        },
        "ref": null,
        "rendered": "Корзина пуста",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "__emotion_base": "h2",
          "__emotion_forwardProp": undefined,
          "__emotion_real": [Circular],
          "__emotion_styles": Array [
            "
    text-align: center;
    font-size: 32px;
    padding: 20px 0;
",
          ],
          "defaultProps": undefined,
          "displayName": "EmptyCartHead",
          "render": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Btn
            title="На главную"
            to="/"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "title": "На главную",
            "to": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "__emotion_base": "div",
          "__emotion_forwardProp": undefined,
          "__emotion_real": [Circular],
          "__emotion_styles": Array [
            "
    text-align:center;
",
          ],
          "defaultProps": undefined,
          "displayName": "Styled(div)",
          "render": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "div",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        "
    max-width:800px;
    padding: 0 10px;
    margin:auto;
",
      ],
      "defaultProps": undefined,
      "displayName": "Styled(div)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef(render)>
            Корзина пуста
          </ForwardRef(render)>,
          <ForwardRef(render)>
            <Btn
              title="На главную"
              to="/"
            />
          </ForwardRef(render)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Корзина пуста",
          },
          "ref": null,
          "rendered": "Корзина пуста",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "__emotion_base": "h2",
            "__emotion_forwardProp": undefined,
            "__emotion_real": [Circular],
            "__emotion_styles": Array [
              "
    text-align: center;
    font-size: 32px;
    padding: 20px 0;
",
            ],
            "defaultProps": undefined,
            "displayName": "EmptyCartHead",
            "render": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Btn
              title="На главную"
              to="/"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "title": "На главную",
              "to": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "__emotion_base": "div",
            "__emotion_forwardProp": undefined,
            "__emotion_real": [Circular],
            "__emotion_styles": Array [
              "
    text-align:center;
",
            ],
            "defaultProps": undefined,
            "displayName": "Styled(div)",
            "render": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
    max-width:800px;
    padding: 0 10px;
    margin:auto;
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
